r_d22/DynamicArray.h

```diff
@@ -0,0 +1,195 @@
+
+        T* newArray = new T[newCapacity];
+        std::size_t elementsToCopy = (newSize < this->size) ? newSize : this->size;
+
+        for (std::size_t i = 0; i < elementsToCopy; ++i) {
+            newArray[i] = this->array[i];
+        }
+        if (newSize > this->size) {
+            for (std::size_t i = elementsToCopy; i < newSize; ++i) {
+                newArray[i] = T();
+            }
+        }
+
+        delete[] this->array;
+
+        this->array = newArray;
+        this->size = newSize;
+        this->capacity = newCapacity;
+    };
+
+    std::size_t getSize() const {
+        return this->size;
+    };
+
+    //Clear internal memory, set size to 0
+    void clear() {
+        delete[] this->array;
+
+        this->size = 0;
+        this->capacity = 0;
+        this->array = nullptr;
+    };
+
+    bool operator==(const DynamicIntArray<T>& other) const {
+        if (this->size != other.getSize()) return false;
+
+        for (std::size_t i = 0; i < this->size; i++) {
+            if (this->array[i] != other[i]) return false;
+        }
+
+        return true;
+    }
+    bool operator!=(const DynamicIntArray<T>& other) const {
+        return !(*this == other);
+    };
+
+    // Allocate new memory for new element
+    // Copy old content to new inner array
+    // insert element at the last index
+    // update size
+    void push_back(T element) {
+        const std::size_t newSize = this->size + 1;
+
+        if (this->capacity < newSize) {
+            std::size_t newCapacity = this->capacity == 0 ? 1 : static_cast<std::size_t>(this->capacity * 1.5);
+
+            T* newArray = new T[newCapacity];
+            for (std::size_t i = 0; i < this->size; ++i) {
+                newArray[i] = this->array[i];
+            }
+            delete[] this->array;
+
+            this->array = newArray;
+            this->capacity = newCapacity;
+        }
+
+        this->array[this->size] = element;
+        this->size = newSize;
+    };
+
+    void reserve(std::size_t reservedSpace) {
+        if (reservedSpace > this->capacity) {
+            T* newArray = new T[reservedSpace];
+
+            for (std::size_t i = 0; i < this->size; i++) {
+                newArray[i] = this->array[i];
+            }
+
+            delete[] this->array;
+            this->array = newArray;
+            this->capacity = reservedSpace;
+        }
+    }
+
+    std::size_t getCapacity() const {
+        return this->capacity;
+    }
+
+    void shrinkToFit() {
+        if (this->capacity > this->size) {
+            T* newArray = new T[this->size];
+
+            for (std::size_t i = 0; i < this->size; i++) {
+                newArray[i] = this->array[i];
+            }
+
+            delete[] this->array;
+            this->array = newArray;
+            this->capacity = this->size;
+        }
+    }
+
+    void pop_back() {
+        if (this->size > 0) {
+            this->size--;
+        }
+
+        this->array[this->size].~T();
```

Comment:
don't call d-tor explicitly.

Reply:
Better change actual capacity from time to time once size is small enough relative to the current capacity and in that way d-tor will be called implicitly as well